[BEAM-10623] Add workflow to run Beam python tests on Linux/Windows/Mac platforms

TobKed opened the pull request:
Workflow run with python tests

# Before merging

### **Before merging it is required to setup related secrets:**
 * `GCP_TESTING_BUCKET` -  beam-github-actions-tests
 * `GCP_PROJECT_ID` -  apache-beam-testing
 * `GCP_REGION` - us-central1

### Before merging it Service Account should have elevated permissions:
 * Storage Admin (`roles/storage.admin`)
 * Dataflow Admin (`roles/dataflow.admin`)

## it was requested in https://issues.apache.org/jira/browse/INFRA-20675

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

TobKed commented:
R: @aaltay @tvalentyn @kamilwu @damgad @potiuk

TobKed commented:
example run executed on my fork (all jobs)

aaltay commented on .github/workflows/python_tests.yml:
```diff
@@ -0,0 +1,196 @@
+
+  build_python_sdk_source:
+    name: 'Build Python SDK Source'
+    if: github.repository_owner == 'apache' && (github.event_name == 'push' || github.event_name == 'schedule')
```
Given the size of this run, maybe we can limit this to 'schedule' only. Do you know how long it takes to run this?

tvalentyn replied:
Current schedule is once a day as per line 22.  How about we run it every 4 hours?

aaltay replied:
Does github impose any limits on how many workflow could run in parallel?

TobKed replied:
Full workflow run takes approximately 30min. check on PR ~15 minutes.
IMHO It is not short, but also not too long. 
If you thinks it is still too long, maybe we could keep schedule and push to master/release branch events only? WDYT?

TobKed replied:
> Does github impose any limits on how many workflow could run in parallel?

Limits for concurrent jobs are between 20-180 depends on the plan ([docs](https://docs.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits)).
I heard that apache has Enterprise plan fot gh-actions but I am not sure about it. Do you have more information about it @gmcdonald ?

aaltay commented on .github/workflows/python_tests.yml:
```diff
@@ -0,0 +1,196 @@
+        run: pip install wheel
+      - name: Build source
+        working-directory: ./sdks/python
+        run: python setup.py sdist --formats=zip
```
does sdist require wheel? If sdist contains binaries, are they compatible with other python version?

TobKed replied:
I checked it and sdist does not require wheel. Source distribution does not contain binaries and is not python/platform specific. I simplified both this job and passing source dist to dataflow job.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_runner_test.py:
```diff
@@ -66,6 +66,7 @@ class InteractiveRunnerTest(unittest.TestCase):
   def setUp(self):
     ie.new_env()
 
+  @unittest.skipIf(sys.platform == "win32", "[BEAM-10627]")
```
@KevinGG - assigned this jira to you.

KevinGG replied:
Acked. It's a duplicate of BEAM-10514. The reason is the file system in test has a temp directory with a very long prefix, causing the file path exceeding windows default 260 character limit.

A normal windows temp directory prefix: `c:\windows\temp\`, 
In the test: `D:\a\beam\beam\sdks\python\target\.tox\py35-win\tmp\`.

The length diff breaks the limit since usually Interactive Beam uses about 240 characters in path length.

TobKed replied:
Thanks @KevinGG . Should I put `BEAM-10514` here or keep `BEAM-10627`?

aaltay commented on sdks/python/apache_beam/runners/portability/portable_runner_test.py:
```diff
@@ -284,6 +284,7 @@ def create_options(self):
     return options
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="[BEAM-10625]")
```
@tvalentyn - - assigned this jira to you.

TobKed replied:
Sure, I added comment about specific environment in the Jira. I will add/extend documentation since https://github.com/apache/beam/pull/12405 is merged.

aaltay commented on sdks/python/apache_beam/runners/portability/portable_runner_test.py:
```diff
@@ -284,6 +284,7 @@ def create_options(self):
     return options
 
 
```
why are we using pytest here insted of unittest? @tvalentyn

aaltay commented on sdks/python/apache_beam/testing/datatype_inference_test.py:
```diff
@@ -174,6 +175,7 @@ def test_infer_typehints_schema(self, _, data, schema):
   @parameterized.expand([(d["name"], d["data"], d["pyarrow_schema"])
                          for d in TEST_DATA])
   @unittest.skipIf(pa is None, "PyArrow is not installed")
+  @unittest.skipIf(sys.platform == "win32", "[BEAM-10624]")
```
@TheNeuralBit - - assigned this jira to you.

TheNeuralBit replied:
Ack I'll take a look. It may just take a pyarrow upgrade, which will be unblocked once we drop python 2

TheNeuralBit replied:
Looks like it's not actually related to pyarrow version. I think we should fix it by making the dtype explicit for the numpy arrays created here: https://github.com/apache/beam/blob/8d03dc698dcea1745a26c9823c8d942f5634c6b1/sdks/python/apache_beam/testing/datatype_inference_test.py#L57

I have a personal Windows machine I could clone Beam on and verify, but it might be simpler to just test the fix as part of this PR.

TobKed replied:
@TheNeuralBit could take a look on https://github.com/apache/beam/pull/12452/commits/4004309b2b32acd1451cd30e92ab14f645a51211 , please?
Is it the solution you mentioned?
I tested it on my local fork (run resulsts [here](https://github.com/TobKed/beam/actions/runs/199604657)) and it seems to work.

TheNeuralBit replied:
Yep that's what I was suggesting!

TobKed replied:
Thanks @TheNeuralBit.  I cherry-picked change to the PR.

tvalentyn commented on .github/workflows/python_tests.yml:
```diff
@@ -0,0 +1,196 @@
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        python: [3.5, 3.6, 3.7, 3.8]
```
Is it possible to extract py versions into a constant on top of the file so that we have to modify them in one place when we add/remove a version?

TobKed replied:
Unfortunately it is not possible right now. At least not in convenient way, especially that in this workflow there are two two type of slightly different matrixes.

TobKed commented on .github/workflows/python_tests.yml:
```diff
@@ -0,0 +1,196 @@
+          # https://github.com/protocolbuffers/protobuf/issues/7765
+          - os: windows-latest
+            python: 3.8
+    steps:
```
Right know it is not possible, however GitHub Actions is evolving so fast it should available soon.
Related links about upcoming related feature:
https://github.com/actions/runner/issues/438
https://github.com/actions/runner/pull/554

TobKed commented on sdks/python/apache_beam/io/parquetio_test.py:
```diff
@@ -296,8 +296,10 @@ def test_sink_transform_int96(self):
   def test_sink_transform(self):
-    with tempfile.NamedTemporaryFile() as dst:
-      path = dst.name
+    fd, path = tempfile.mkstemp()
```
There was problem on windows since `with tempfile.NamedTemporaryFile() as dst` provide `tempfile` which is already opened. On windows any opened file seems to be read only which coused PermissionError. I know this workaround is ugly but it the best way I found to do it.
Is separate commit with JIRA will not be lost after merging?

tvalentyn replied:
You can use  `tempfile.TemporaryDirectory()` to generate a fresh temp directory and use any name as a file name within this directory with the test.

TobKed replied:
I updated it with `tempfile.TemporaryDirectory()`, looks much more better and intuitive now. Thanks a lot!

TobKed commented on sdks/python/apache_beam/runners/worker/log_handler_test.py:
```diff
@@ -87,7 +87,8 @@ def _verify_fn_log_handler(self, num_log_entries):
         self.assertTrue(
-            re.match(r'.*/log_handler_test.py:\d+', log_entry.log_location),
+            re.match(
+                r'.*(/|\\)log_handler_test.py:\d+', log_entry.log_location),
```
Added fix. Thanks.

TobKed replied:
I used `rf'.*{os.sep}log_handler_test.py:\d+'` but it caused:
```
2020-08-05T14:25:17.1905260Z >                   raise source.error("bad escape %s" % escape, len(escape))
2020-08-05T14:25:17.1905424Z E                   sre_constants.error: bad escape \l at position 2
```
https://github.com/apache/beam/pull/12452/checks?check_run_id=949595653
I reverted change.

Should I use it somehow differently?

tvalentyn replied:
You could use `re.escape()`. Not a big deal since this is a test, feel free to leave as is or simplify to `r'.*log_handler_test.py:\d+'`.

TobKed replied:
I had problems with `re.escape()` as well (escaped `\d` as well). For the most clean solution I just simplified it as you suggested it.

TobKed commented on sdks/python/apache_beam/typehints/typecheck_test_py3.py:
```diff
@@ -92,23 +93,29 @@ def test_wrapper_pass_through(self):
+
+    fd, path = tempfile.mkstemp()
+    try:
+      os.close(fd)
```
WDYT about adding comment in such places with JIRA ?

tvalentyn replied:
The reference to JIRA in #12452 (comment) was that you could make test changes in a separate commit such as:

`[BEAM-123456] Fix tests that don't pass on windows` 

and have another commit:

`[BEAM-10623] Add workflow to run Beam python tests on Linux/Windows/Mac platforms`.

TobKed commented on .github/workflows/python_tests.yml:
```diff
@@ -0,0 +1,196 @@
+
+
+env:
+  GCP_WHEELS_STAGING_PATH: "gs://${{ secrets.GCP_BUCKET }}/${GITHUB_REF##*/}/${GITHUB_SHA}-${GITHUB_RUN_ID}/"
```
It is unnecessary copy-paste leftover. Deleted.

TobKed commented:
@tvalentyn answering your question:

> Also, all actions workflow run in parallel, right? so it takes roughly the same time as running a regular precommit? In such case we could consider running it on PR.

Building sdk source dist is used as prerequisite for `python_wordcount_dataflow` job . Building source distribution is already validated for PR in `.github/workflows/build_wheels.yml` workflow.

TobKed commented:
@tvalentyn thanks for review.

> * Is there a way how an author or a reviewer can rerun the action to retry the test or the only way is to re-push the PR?

There is an event type called `workflow_dispatch` which enables button to manually trigger workflow I just added it.
https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/

> * When we run unit tests, we create xml files with test results (see `generated xml file: D:\a\beam\beam\sdks\python\pytest_py36-win.xml ` entries in console logs). It would be great if we could somehow visualize that to see failing tests/messages easily without having to dig through console logs. Are there any options to do that? This could arrive in future PRs.

I am not sure about visualisation, I have to investigate it more depth, however I add uploads of xml files as artifacts.

tvalentyn commented:
> @tvalentyn answering your question:
> 
> > Also, all actions workflow run in parallel, right? so it takes roughly the same time as running a regular precommit? In such case we could consider running it on PR.
> 
> Building sdk source dist is used as prerequisite for `python_wordcount_dataflow` job . Building source distribution is already validated for PR in `.github/workflows/build_wheels.yml` workflow.

Not sure I follow. I meant to say: if Github Action test suites run comparative amount of time as Jenkins precommit suites, we can continue running Github Action test suites on each PR, assuming Actions Quota is not a concern. I wanted to know whether  different suites from the matrix (Py3.5 on Windows, Py3.6 on MacOS) run in parallel or not.

TobKed commented:
> > @tvalentyn answering your question:
> > > Also, all actions workflow run in parallel, right? so it takes roughly the same time as running a regular precommit? In such case we could consider running it on PR.
> > 
> > 
> > Building sdk source dist is used as prerequisite for `python_wordcount_dataflow` job . Building source distribution is already validated for PR in `.github/workflows/build_wheels.yml` workflow.
> 
> Not sure I follow. I meant to say: if Github Action test suites run comparative amount of time as Jenkins precommit suites, we can continue running Github Action test suites on each PR, assuming Actions Quota is not a concern. I wanted to know whether different suites from the matrix (Py3.5 on Windows, Py3.6 on MacOS) run in parallel or not.

Yes, they run in parallel. It is not always the case since parallel jobs are limited to some amount, but AFAIK since apache has Enteprise plan so this number is a quite big ([github actions docs - usage limits](https://docs.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits))

TobKed commented:
I rebased on the latest master, added documentation and badges, applied `TemporaryDirectory` context manager for python2.7 backward tests compatibility.

tvalentyn commented on CI.md:
```diff
@@ -75,8 +75,28 @@ run categories. Here is a summary of the run categories with regards of the jobs
+Some of the jobs require variables stored as a [GitHub Secrets](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets)
+to perform operations on Google Cloud Platform. Currently these jobs are limited to Apache repository only.
+These variables are:
+ * `GCP_PROJECT_ID` - ID of the Google Cloud project. Apache/Beam repository has it set to `apache-beam-testing`.
```
For my education, which repository we refer to here? Also, I would not hard code existing values in documentation, since the source-of-truth may be updated, and documentation may not. You could say: `For example apache-beam-testing` instead.

TobKed replied:
I agree. I changed it referring values as examples.

tvalentyn commented on CI.md:
```diff
@@ -75,8 +75,28 @@ run categories. Here is a summary of the run categories with regards of the jobs
 
+### Google Cloud Platform Credentials
+
+Some of the jobs require variables stored as a [GitHub Secrets](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets)
```
nit: as [GitHub Secrets]

TobKed replied:
Fixed.

tvalentyn commented on sdks/python/apache_beam/testing/util.py:
```diff
@@ -334,3 +335,19 @@ def open_shards(glob_pattern, mode='rt', encoding='utf-8'):
   return io.open(concatenated_file_name, mode, encoding=encoding)
+
+
+class TemporaryDirectory(object):
```
Thanks. I forgot we are still on Py2, but very soon we can start removing Py2 support.

We actually defined this at least twice already. 

https://github.com/apache/beam/blob/0098eb6f9900849b05bfc2ec2497ed8c7d37148a/sdks/python/apache_beam/utils/subprocess_server_test.py#L39

https://github.com/apache/beam/blob/0098eb6f9900849b05bfc2ec2497ed8c7d37148a/sdks/python/apache_beam/testing/test_utils.py#L43

Feel free to reuse one of them and add a comment # TODO(BEAM-8371): Use tempfile.TemporaryDirectory.

TobKed replied:
Thank you. I was not aware of them. I used this first one since it returns path on enter.

TobKed commented on CI.md:
```diff
@@ -75,8 +75,28 @@ run categories. Here is a summary of the run categories with regards of the jobs
+
 | Job                                             | Description                                                                                                                                                                                                                                                        | Pull Request Run | Direct Push/Merge Run | Scheduled Run | Requires GCP Credentials |
 |-------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------|-----------------------|---------------|--------------------------|
+| Check are GCP variables set                     | Checks are GCP variables are set. Jobs which required them depends on the output of this job.                                                                                                                                                                      | Yes              | Yes                   | Yes           | Yes/No                   |
```
I used `Check GCP variables` and also changed it in workflow itself. Looks much more cleaner and simpler.

TobKed commented on CI.md:
```diff
@@ -85,16 +105,15 @@ Those jobs often have matrix run strategy which runs several different variation
- * Storage Object Admin (roles/storage.objectAdmin)
+| Job                              | Description                                                                                                           | Pull Request Run | Direct Push/Merge Run | Scheduled Run | Requires GCP Credentials |
+|----------------------------------|-----------------------------------------------------------------------------------------------------------------------|------------------|-----------------------|---------------|--------------------------|
+| Check are GCP variables set      | Checks are GCP variables are set. Jobs which required them depends on the output of this job.                         | Yes              | Yes                   | Yes           | Yes/No                   |
```
Fixed.

tvalentyn commented:
Thanks, @TobKed , I don't have additional comments. Once all reviewers, whose feedback you expect, gave their LGTM, please squash fixup commits, and leave only the commits that you want to be added to the commit history.

TobKed commented:
> Thanks, @TobKed , I don't have additional comments. Once all reviewers, whose feedback you expect, give their LGTM, please squash fixup commits, and leave only the commits that you want to be added to the commit history.

I found I missed to use `TemporaryDirectory()` in `sdks/python/apache_beam/typehints/typecheck_test_py3.py`, I fixed it.
I also added input to manual trigger so dataflow tests have to be explicitly enabled (made them disabled by default).

Thanks for reviews @tvalentyn and all people involved in this PR. I will create PR to Apache Infra with request for setting up proper secrets.

TobKed commented:
Run Python PreCommit

TobKed commented:
Run Python2_PVR_Flink PreCommit

TobKed commented:
Run Python PreCommit

TobKed commented:
Created single PR to infra with all required changes for github secrets: https://issues.apache.org/jira/browse/INFRA-20675

tvalentyn commented:
> I also added input to manual trigger so dataflow tests have to be explicitly enabled (made them disabled by default).
Are Dataflow tests enabled in automated execution or not?

TobKed commented:
> > I also added input to manual trigger so dataflow tests have to be explicitly enabled (made them disabled by default).
> > Are Dataflow tests enabled in automated execution or not?

Dataflow tests for automated execution are enabled as they before (events: push/merge, scheduled). Behavior changed  for manual trigger. When triggered by manual button they may be able do choose are dataflow tests shall run or not. .

TobKed commented:
I rebased on the latest master and solved conflicts.

tvalentyn commented:
Thanks, looks like 'Upload test logs' is flaky:
```
 Upload test logs1s
    pythonLocation: C:\hostedtoolcache\windows\Python\3.5.4\x64
Run actions/upload-artifact@v2
  with:
    name: pytest-windows-latest-3.5
    path: sdks/python/pytest**.xml
    if-no-files-found: warn
  env:
    pythonLocation: C:\hostedtoolcache\windows\Python\3.5.4\x64
With the provided path, there will be 2 file(s) uploaded
##[error]read ECONNRESET
```

```
Exponential backoff for retry #1. Waiting for 5006.55256840354 milliseconds before continuing the upload at offset 0
##[error]Unexpected response. Unable to upload chunk to https://pipelines.actions.githubusercontent.com/COKSSHEuHyOLM1pP93G8aBnLpu5tMJfYiveSzILlm0cIfDvSEl/_apis/resources/Containers/4328416?itemPath=pytest-ubuntu-latest-3.8%2Fpytest_py38_no_xdist.xml
##### Begin Diagnostic HTTP information #####
Status Code: 404
Status Message: Not Found
Header Information: {
  "content-length": "29",
  "content-type": "text/plain",
  "strict-transport-security": "max-age=2592000",
  "x-tfs-processid": "98f88b46-bb0d-4ece-af51-fbf9accea451",
  "x-tfs-serviceerror": "The+resource+cannot+be+found.",
  "x-msedge-ref": "Ref A: E85C8E28FC70417687E1D2AAE3FBDD17 Ref B: BN3EDGE0515 Ref C: 2020-08-17T18:55:44Z",
  "date": "Mon, 17 Aug 2020 18:55:53 GMT"
}
###### End Diagnostic HTTP information ######
##[warning]Aborting upload for /home/runner/work/beam/beam/sdks/python/pytest_py38_no_xdist.xml due to failure
  Post Checkout code
```
Is there a way we can configure additional retries there?

Also, how a PR author will re-run the actions tests on their PR? Is it possible to rerun only failed tests? (I remember I asked this question before, but I forgot what the answer was)

tvalentyn commented:
> Also, how a PR author will re-run the actions tests on their PR? Is it possible to rerun only failed tests? (I remember I asked this question before, but I forgot what the answer was)

Found your reply above and also found a `Rerun all jobs` button, which I clicked on one of the failed jobs.

TobKed commented:
I rebased on the latest master and solved conflicts in CI.md.

@tvalentyn regarding flaky upload I tested it so many times and it never happened before. IMHO it was a temporary problem with GitHub Actions itself. 
Related issue https://github.com/actions/upload-artifact/issues/116

TobKed commented:
@tvalentyn @aaltay I see that https://issues.apache.org/jira/browse/INFRA-20675 is done and just waits for confirmation.

tvalentyn commented:
Ack, thanks. Is this ready to merge?

TobKed commented:
@tvalentyn yes

TobKed commented:
Thank you!
